### DISCUSSION TITLE
How can I caught ERR_HTTP_HEADERS_SENT

### DISCUSSION THREAD

**Discussion by yasmanets** [nodejs/help#4749] [Q&A]
### Version
`16.15`

### Problematic

I have a server that exposes an endpoint that responds with a file through a `pipeline`.  Cancelling the request results in an `ERR_STREAM_PREMATURE_CLOSE`. 

When that error occurs, an `ERR_HTTP_HEADERS_SENT` is thrown but I am not able to catch it.

I have tried catching it by listening to `process.on('uncaughtException')` and `request.on('error')` but both options don't work.

Any suggestions or ideas about the problem?

**Comment by atlowChemi** [comment]
Basically you would need to add a listener to the `close` event of the request, and when the `ERR_STREAM_PREMATURE_CLOSE` is detected, you could destroy file's stream:

```javascript
const http = require('node:http');

const server = http.createServer((req, res) => {
  const responseStream = yourFileProcessingPipeline(); // Your file processing pipeline here

  // Listen for premature close event on the response
  req.on('close', () => {
    // Abort the responseStream when the premature close occurs
    responseStream.destroy(new Error('Premature close occurred'));
  });

  responseStream.on('error', (err) => {
    if (err.code === 'ERR_HTTP_HEADERS_SENT') {
      // Handle ERR_HTTP_HEADERS_SENT error here
      console.error('Headers already sent:', err.message);
    } else {
      // Handle other errors that occurred during the response pipeline
      console.error('An error occurred:', err.message);
    }
  });

  responseStream.pipe(res); // Pipe the response stream to the HTTP response
});

server.listen(8080, () => {
  console.log('Server is running on port 8080');
});
```
